File: Behavioral/Delegation/kotlin/delegation.kt
interface ISoundBehaviour {
    fun makeSound()
}

class CatSound : ISoundBehaviour{
    override fun makeSound() {
        println("Meow")
    }
}

class CatRelaxSound : ISoundBehaviour{
    override fun makeSound() {
        println("Prrrrrr")
    }
}


class Cat {
    var sound : ISoundBehaviour = CatSound()

    fun makeSound(){
        this.sound.makeSound()
    }

    fun setSoundBehaviour(newSound : ISoundBehaviour){
        this.sound = newSound
    }
}

fun main(args: Array<String>) {
    val cat : Cat = Cat()
    cat.makeSound() // Meow
    val newSound : ISoundBehaviour = CatRelaxSound()
    cat.setSoundBehaviour(newSound)
    cat.makeSound() // Prrrr
}

File: Behavioral/Delegation/typescript/delegation.ts
interface ISoundBehaviour {
  makeSound: () => void;
}

class CatSound implements ISoundBehaviour {
  makeSound() {
    console.log("Meow");
  }
}

class CatRelaxSound implements ISoundBehaviour {
  makeSound() {
    console.log("Prrrrrr");
  }
}

class Cat {
  sound: ISoundBehaviour;
  constructor() {
    this.sound = new CatSound();
  }

  makeSound() {
    this.sound.makeSound();
  }

  setSoundBehaviour(newSound: ISoundBehaviour) {
    this.sound = newSound;
  }
}

const cat = new Cat();
cat.makeSound(); // Meow
const newSound: ISoundBehaviour = new CatRelaxSound();
cat.setSoundBehaviour(newSound);
cat.makeSound(); // Prrrr
